design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,315 @@
+
+### Design
+
+This feature adds an additional controller for the CRD `Volume Migration` to virt-controller. Once the VolumeMigration object is created then KubeVirt will update the VMI status with the source and destination volume pair.
+
+Afterwards, the controller will create a VirtualMachineInstanceMigration to trigger the VM live migration.
+
+If there are some migrated volumes in the VMI status, the VM migration controller will replace the destination volumes by creating the target virt-launcher pod with the new volumes and virt-launcher will select the source volume to be migrated.
+
+Once the migration completes, the controller will update the migrated volume first in the VMI spec. This will trigger the update of the volume in the VM spec if it exists.
+
+DataVolumes are taken into account in the current proposal. DataVolumes are used to import the image during VM provisioning, but they may essentially be thought of as PVCs for the rest of the VM life cycle. According to this proposal, KubeVirt will replace the source DataVolumes with the destination PVC when a user wishes to migrate data to a PVC.
+
+A schematic representation of the migration flow is shown below.
+![](volume-migration.png)
+
+### API examples
+
+Example of VolumeMigration
+
+Create a VM with source storage and the destination PVC:
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: dest-pvc
+spec:
+  accessModes:
+    - ReadWriteOnce
+  volumeMode: Filesystem
+  resources:
+    requests:
+      storage: 1Gi
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  annotations:
+    kubevirt.io/libvirt-log-filters: "1:qemu.qemu_monitor 1:*"
+  labels:
+    kubevirt.io/vm: vm-alpine-datavolume
+  name: vm-alpine-datavolume
+spec:
+  dataVolumeTemplates:
+  - metadata:
+      creationTimestamp: null
+      name: source-pvc
+    spec:
+      pvc:
+        accessModes:
+        - ReadWriteOnce
+        resources:
+          requests:
+            storage: 1Gi
+        storageClassName: local
+      source:
+        registry:
+          url: docker://registry:5000/kubevirt/alpine-container-disk-demo:devel
+  running: true
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-alpine-datavolume
+    spec:
+      domain:
+        devices:
+          disks:
+          - disk:
+              bus: virtio
+            name: datavolumedisk1
+          interfaces:
+          - masquerade: {}
+            name: default
+        resources:
+          requests:
+            memory: 128Mi
+      networks:
+      - name: default
+        pod: {}
+      terminationGracePeriodSeconds: 0
+      volumes:
+      - dataVolume:
+          name: source-pvc
+        name: datavolumedisk1
+
+```
+
+Trigger the migration with:
+```yaml
+apiVersion: storage.kubevirt.io/v1alpha1
+kind: VolumeMigration
+metadata:
+  name: sc-mig
+spec:
+  vmiName: vmi
```

Comment:
Currently, there is no `vmiName` field in the `VolumeMigration` definition. Shall we add it to explicitly state that we support only a single VMI per request?

Reply:
sorry, this is a leftover and it should be removed